modal-go/cls.go

```diff
@@ -0,0 +1,215 @@
+package modal
+
+// Cls lookups and Function binding.
+
+import (
+	"context"
+	"fmt"
+
+	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
+	"google.golang.org/protobuf/proto"
+)
+
+type Cls struct {
+	ClassId           string
+	methods           map[string]*Function
+	schema            []*pb.ClassParameterSpec // schema for parameters used in init
+	serviceFunctionId string
+	ctx               context.Context
+}
+
+// Creates parametrized instnace of Cls.
+func (c *Cls) Instance(kwargs map[string]any) (*Cls, error) {
+	if len(c.schema) == 0 {
+		return c, nil
+	} else {
+
+		// Create a parameter set with values from kwargs
+		paramSet := pb.ClassParameterSet_builder{
+			Parameters: []*pb.ClassParameterValue{},
+		}.Build()
+
+		// For each parameter in the schema
+		for _, paramSpec := range c.schema {
+			name := paramSpec.GetName()
+			value, provided := kwargs[name]
+
+			// Skip if not provided and has default value
+			if !provided && paramSpec.GetHasDefault() {
+				continue
+			}
+
+			// Error if required and not provided
+			if !provided && !paramSpec.GetHasDefault() {
+				return nil, fmt.Errorf("required parameter '%s' not provided", name)
+			}
+
+			// Convert value to proto parameter
+			paramValue, err := convertToParameterValue(name, paramSpec.GetType(), value)
+			if err != nil {
+				return nil, err
+			}
+
+			newParameters := append(paramSet.GetParameters(), paramValue)
+			paramSet.SetParameters(newParameters)
+		}
+
+		// Serialize the parameter set
+		serializedParams, err := proto.Marshal(paramSet)
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize parameters: %w", err)
+		}
+
+		// Bind parameters to create a parameterized function
+		bindResp, err := client.FunctionBindParams(c.ctx, pb.FunctionBindParamsRequest_builder{
+			FunctionId:       c.serviceFunctionId,
+			SerializedParams: serializedParams,
+		}.Build())
+		if err != nil {
+			return nil, fmt.Errorf("failed to bind parameters: %w", err)
+		}
+
+		// Update all methods to use the bound function ID
+		boundFunctionId := bindResp.GetBoundFunctionId()
+		for methodName, function := range c.methods {
+			function.FunctionId = boundFunctionId
+			c.methods[methodName] = function
+		}
+	}
+	return c, nil
+}
+
+// Method returns the Function with the given name.
+// It returns an error if the Cls is not initialized or if the method doesn't exist.
+func (c *Cls) Method(name string) (*Function, error) {
```

Comment:
IMO we shouldn't allow getting methods directly on the Cls, and always have the user go via instantiation, even for non-parameterized classes. This is a change we recently made in the Python SDK as well (you used to be able to do `SomeClass.some_method.remote()` without instantiation) in order to get a more consistent API (will allow us to put certain logic we expect to run for all classes in the instantiation logic)